### DISCUSSION TITLE
How do you make those plugin docs?

### DISCUSSION THREAD

**Discussion by Andre601** [lowlighter/metrics#1312] [Q&A] [answered]
I'm always amazed at the design and look of the docs for the different plugins.

Looking at them, its clear for me that you're using a tool/software to generate them automatically.
My question now is, is it a selfmade one? If so, is it separately available? Also, can it be used for other stuff too?

I want to update a project of mine and would like to use such a tool to cut out repeated information or similar.

**Comment by lowlighter** [comment, marked as answer]
Thanks for appreciating the docs 😄 !

Documentation files are mostly auto-generated from a selfmade tool but unfortunately the code for it is a bit scattered everywhere in the codebase so it'll be hard to reuse it elsewhere, and it's tightly coupled for *metrics* use case anyway so I don't think it's easily reusable as it is currently... It's also not something I plan to refactor (at least in the near future).

So unfortunately it's highly improbable you'll be able to reuse the same code as *metrics*.
But I can explain briefly how it works though if you're interested about it, maybe it can give you ideas 🙂 
___

The [`.github/scripts/build.mjs`](https://github.com/lowlighter/metrics/blob/master/.github/scripts/build.mjs) script is the entrypoint of all generated files (not necessarily documentation, the workflows too are auto-generated for example).

The aim of *metrics* documentation is to follow the *"The truth is in the code"* principle by having:
- code ≈ documentation
  - `metadata.yml` documents available options, but also serves as inputs parsing/casting/validation/defaulting code-wise
- tests ≈ examples
  - `examples.yml` displays examples, but also serves as integration tests in the CI/CD pipeline

While the docs from [`.github/readme/partials`](https://github.com/lowlighter/metrics/tree/master/.github/readme/partials) are templated with [EJS](https://github.com/mde/ejs), the plugins documentation tables are generated in `source/app/metrics/metadata.mjs` [here for plugins](https://github.com/lowlighter/metrics/blob/73c8e87dc7bb8f874eef1585add26cfef81fbd0a/source/app/metrics/metadata.mjs#L405-L534) and [here for templates](https://github.com/lowlighter/metrics/blob/73c8e87dc7bb8f874eef1585add26cfef81fbd0a/source/app/metrics/metadata.mjs#L563-L598) (but beware, it looks like weird JSX with HTML strings within JS, not the code I'm the most proud of 😅 ). Also, the code itself relies heavily on the structure of the project, everything under `source/plugins` is expected to be a plugin implementation and same for `source/templates` which expects templates definition (it uses fileglobs to iterate over existing components).

If you look at a plugin `metadata.yml` you'll notice that `README.md` is actually just a nice formatting over it. The table looks nice because they use HTML tables with rowspan/colspan and cool formatting, but the inconvenient is that it's not really markdown anymore (but it does looks nice on GitHub). The templating itself is done by replacing the content of special HTML comments (which are hidden in markdown):

```markdown
<!--header-->
<!--/header-->

## ➡️ Available options

<!--options-->
<!--/options-->

... Some static documentation that can be written by plugin authors...

## ℹ️ Examples workflows

<!--examples-->
<!--/examples-->
```

https://github.com/lowlighter/metrics/blob/5100ebadfe1c4b43fcdceaf9ec601385896b67a4/.github/scripts/build.mjs#L41-L44

It's not a perfect solution and the implementation is kind of inconsistant (some templating done in EJS and some in JS) but at least it leverage the amount of work needed for maintenance, and the documentation is guaranteed to be up-to-date and reflects the code while the examples are supposed to be working too since they're tested on each build

**Reply by Andre601** [reply]
I kinda expected this to be the case (selfmade tool mostly included in metrics itself).

I manually recreated it myself now. Took a few hours but I feel it was worth it.

https://github.com/Andre601/Formatter-Expansion

**Reply by lowlighter** [reply]
Looks nice indeed !
